### ISSUE TITLE
Default imports not imported properly

### ISSUE THREAD

**Issue by yajium** [withastro/astro#6702]
### What version of `astro` are you using?

1.22.19

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

yarn

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

This issue seems to remain unresolved. In the latest version of the project, when importing a module with both default and named exports, and then using the function exported as the default, an error occurs only during the build process.

## Error Messsage
error　markdownToHtml is not a function

##  Steps to Reproduce
import default exported function like this:
```ts
import markdownToHtml from "zenn-markdown-html";
```

the content of "zenn-markdown-html" file is this:
```ts
import { MarkdownOptions } from './types';
declare const markdownToHtml: (text: string, options?: MarkdownOptions) => string;
export default markdownToHtml;
export { markdownToSimpleHtml } from './markdown-to-simple-html';
```

and I tried checking how this function is outputted in `astro build`, and then result is like this:
```ts
{ default: [Function: markdownToHtml], markdownToSimpleHtml: [Getter] }
```

so I have to use this funciton as `markdownToHtml.default`, but this method results in red squiggly error lines in VSCode, which is a bit unpleasant. Could you please investigate again?

## Details
zenn-markdown-html:
https://github.com/zenn-dev/zenn-editor/tree/canary/packages/zenn-markdown-html

### Link to Minimal Reproducible Example

https://github.com/yajium/astro-export-test

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by bluwy** [comment]
This is because `zenn-markdown-html` isn't exported properly. If you check the exported code at https://www.npmjs.com/package/zenn-markdown-html?activeTab=code `lib/index.js`, it's exporting the `markdownToHtml` at `exports.default`. This is the behaviour you see in build so it errors out.

In dev, it's not this way because Vite respects the `exports.__esModule` hint (you can see it's assigned at the top of the file), so `.default` becomes the "default". Honestly I think Vite shouldn't respect the `__esModule` for external packages to ensure consistency, but that needs to be done in the next major.

For now, you should stick to named imports to avoid this issue. Or request the library to export a node-default-import-friendly API. Closing as there's not much Astro can do for now.

**Comment by ymeine** [comment]
I apologize in advance for piggybacking on a closed issue, but I faced the exact same one for Iconify's offline React icons.

I am having trouble understanding a couple of things: 

- what is the reason the build process is so different to the dev one? (in *many* other ways too, but that would deviate from this specific issue)
  - is the build not using Vite?
  - how are modules resolved, processed and loaded in build vs dev?
- what is the "node-default-import-friendly API" you mention? Is it a matter of changing: their file content (built code)? their file extension? their package.json? A mix of this?

As a consumer of libraries and tools, I'm too frequently facing issues linked to this schizophrenia of module systems. At least it could help if someone with higher knowledge of the system could explain the differences between build process and dev process, so one can deduce potential implications.

Thanks in advance

**Comment by bluwy** [comment]
> * what is the reason the build process is so different to the dev one? (in _many_ other ways too, but that would deviate from this specific issue)

I think this is a bug in Vite dev, where it would handle `__esModule` by default so things work as expected, but it doesn't reflect the build behaviour. I've noted this breaking change for Vite 5 somewhere to fix it.



> * is the build not using Vite?
> * how are modules resolved, processed and loaded in build vs dev?

It is using Vite, but when building a bundle, there's two ways of handling your dependencies:

1. Bundle them together
2. Don't bundle them (aka externalize), so the bundle refers with a raw `import "my-dep"` in the bundled code.

no2 is the culprit here as the `import "my-dep"` runtime is handled by Node, and not Vite. And Node doesn't recognize `__esModule` where the problem exist.

> * what is the "node-default-import-friendly API" you mention? Is it a matter of changing: their file content (built code)? their file extension? their package.json? A mix of this?

Since Node doesn't understand `__esModule`, you'd have to awkwardly use the `.default` property of the package to work, which is the not-so-friendly part. Ideally the package should use `export default 'foo'` instead of `export default { default: 'foo' }` (or similar indirect pattern).

**Comment by ymeine** [comment]
> I think this is a bug in Vite dev, where it would handle __esModule by default so things work as expected, but it doesn't reflect the build behaviour. I've noted this breaking change for Vite 5 somewhere to fix it.

Thanks for sharing that piece of information.

> Don't bundle them (aka externalize), so the bundle refers with a raw import "my-dep" in the bundled code.
> no2 is the culprit here as the import "my-dep" runtime is handled by Node, and not Vite. And Node doesn't recognize __esModule where the problem exist.

So here's what I saw on my side: 

- DEV mode: 
  - Node.js side: `__esModule` handled
  - client side: `__esModule` handled
- build mode: 
  - Node.js side: `__esModule` **NOT** handled
  - client side: `__esModule` handled

For the client part it's consistently handling it and it works, as expected. For the Node.js side we see the *"favorable"* bug you mentioned previously in DEV mode. Adding to that what you said next, I assume dependencies are all externalized on the Node.js side.

> Since Node doesn't understand __esModule, you'd have to awkwardly use the .default property of the package to work, which is the not-so-friendly part. Ideally the package should use export default 'foo' instead of export default { default: 'foo' } (or similar indirect pattern).

I was more referring to how 3rd parties should bundle their code and distribute their packages to accommodate all situations. But even if they do, I think there might still be some issues.

As an example I faced too: [octokit](https://github.com/octokit/octokit.js). If you check https://www.unpkg.com/octokit@2.0.21/package.json you will see: 

```json
{
  "main": "dist-node/index.js",
  "module": "dist-web/index.js",
}
```

So many things to tell here: 

- `main` is an *official* field defined by npm, and here is associated to a Node.js friendly bundle, but we don't know if it's ESM or CommonJS
- `module` isn't defined by npm, so I guess Rollup or one of its plugins *is able* to consume it (but I don't know whether or when they do). The associated path suggests it is a bundle adapted for use on the client side. But again, nothing tells if it's ESM or CommonJS, though I assumed "module" implied ESM.

We're starting to see a combination of possibilities, and I think the current spec of `package.json` is way too limited to accommodate all use cases. Each potential target should have an entry (browser, Node.js, etc.) and for each, theye should be one sub entry per bundle type (ESM, CommonJS, etc.).

I eventually couldn't use that dependency, since `main` (`dist-node`) was used on both Node.js side and client side, and I couldn't tell Astro, or Vite, or Rollup I wanted to use different versions depending on the context. Even with an explicit import to the desired bundle, I could not handle indirect dependencies, and they use a lot of internal ones distributed the same way.

----

I know it derives a bit from the initial topic of that GitHub issue, but it adds clarity to the question I asked in my previous comment.

Eventually, I can say the diversity of module systems (at least ESM and CommonJS), coupled to the diversity of platforms JavaScript targets (in Astro there are two: Node.js and browser) makes the whole problem very complex, and sometimes impossible to solve.

**Comment by bluwy** [comment]
Yep you pretty much summarised the state of JS 😄 There's also the `"exports"` field that helps with CJS/ESM browser/nodejs/etc matrices when exporting things. New packages should use the field if possible, older packages would have to cut a breaking release if they use it, so not many of them have adopted it yet. I've also made https://publint.dev to help with this.

**Comment by ymeine** [comment]
> Yep you pretty much summarised the state of JS

Well, I realize that what I described may actually be a good candidate for releasing multiple packages instead, one per combination. That could be a fine tradeoff.

Thanks for sharing https://publint.dev/, I'm bookmarking it.
